frame/contracts/src/lib.rs

```diff
@@ -1052,9 +1065,15 @@ struct CallInput<T: Config> {
 	determinism: Determinism,
 }
 
+/// Reference to an existing code hash or a new wasm module.
+enum WasmCode<T: Config> {
```

Comment:
This one is used by the API, we can't really get rid of it without making breaking API changes. 
If we want to have API functions that can operate with either the code or a hash, we can't get away without having this type.

Now, once you get into the body of `instantiate_with_code`, if you have the `Upload<Vec<u8>>` variant, you are going to take this input and transform it into a `WasmBlob` so that you can store the code, and later execute it.

When we run the code we reuse this `WasmBlob` or get it from storage with the hash
https://github.com/paritytech/substrate/blob/96e3dcf9a77fa33c253c2bd1a77dba161361cc89/frame/contracts/src/lib.rs#L1253-L1256

Before this refactoring, we were doing this conversion at this stage, when we execute the code, and then we would store the code on chain if needed. This PR changes that so that the code is always in storage first so that you can rely on the same  storage state whether you call instantiate or instantiate_with_code